src/rust/ensogl/lib/components/src/file_browser/column.rs

```diff
@@ -0,0 +1,442 @@
+        icon.deref().set_position_x(ENTRY_PADDING_LEFT+icons::ICON_SIZE /2.0);
+
+        Self { display_object,label,icon,arrow }
+    }
+
+    fn width(&self) -> f32 {
+        ENTRY_PADDING_LEFT + icons::ICON_SIZE + ICON_TO_LABEL + self.label.width.value()
+            + LABEL_TO_ARROW + icons::ICON_SIZE + ENTRY_PADDING_RIGHT
+    }
+}
+
+impl display::Object for ListEntry {
+    fn display_object(&self) -> &display::object::Instance {
+        &self.display_object
+    }
+}
+
+impl list_view::entry::Entry for ListEntry {
+    fn set_focused(&self, focused: bool) {
+        let text_color = if focused {
+            entry_color_focused()
+        } else {
+            entry_color_normal()
+        };
+        self.label.set_color_all(text_color);
+        self.icon.set_focused(focused);
+        if let Some(arrow) = &self.arrow {
+            arrow.set_focused(focused);
+        }
+    }
+
+    fn set_width(&self, width: f32) {
+        if let Some(arrow) = &self.arrow {
+            arrow.set_position_x(width-ENTRY_PADDING_RIGHT-icons::ICON_SIZE /2.0);
+        }
+    }
+}
+
+#[derive(Debug,Clone,CloneRef,Default)]
+struct ListEntryProvider(Rc<Vec<Rc<ListEntry>>>);
+
+impl ListEntryProvider {
+    fn new<'a>(app:&Application, entries:impl IntoIterator<Item=&'a Entry>) -> Self {
+        let entries = entries.into_iter().map(|entry| Rc::new(ListEntry::new(app,entry.clone())));
+        ListEntryProvider(Rc::new(entries.collect_vec()))
+    }
+
+    fn width(&self) -> f32 {
+        let widths = self.0.iter().map(|entry| entry.width());
+        widths.max_by(|x,y| x.partial_cmp(y).unwrap()).unwrap_or(0.0)
+    }
+}
+
+impl list_view::entry::EntryProvider for ListEntryProvider {
+    fn entry_count(&self) -> usize {
+        self.0.len()
+    }
+
+    fn get(&self, _app: &Application, id: usize) -> Option<list_view::entry::AnyEntry> {
+        Some(list_view::entry::AnyEntry::from(self.0.get(id)?.clone()))
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Clone,Debug)]
+pub struct Model {
+    app            : Application,
+    pub entries    : RefCell<Option<Rc<Vec<Entry>>>>,
+    pub list_view  : ListView,
+    state_label    : text::Area,
+    shadow         : column_shadow::View,
+    display_object : display::object::Instance,
+}
+
+
+
+// ===========
+// === FRP ===
+// ===========
+
+ensogl_core::define_endpoints! {
+    Input {
+        set_entries (Rc<Vec<Entry>>),
+        set_error   (ImString),
+    }
+    Output {
+        entry_selected (PathBuf),
+        entry_chosen   (PathBuf),
+        left           (f32),
+        right          (f32),
+        width          (f32),
+    }
+}
+
+
+
+// ==============
+// === Column ===
+// ==============
+
+#[derive(Debug,Clone,CloneRef)]
+pub struct Column {
+    pub model : Rc<Model>,
+    frp       : Frp,
+}
+
+impl Column {
+    pub fn new(browser:Rc<ModelWithFrp>, index:usize) -> Column {
+        let weak_browser   = Rc::downgrade(&browser);
```

Comment:
I don't see good way to implement this without any weak pointers at all. The core issue is that events in the column's network can lead to the creation of new columns, which requires a reference to the file browser's network which has references to the browser's `Model` which has a reference to the column. This forms a cycle.

I think the difference to nodes is that they do not fire any events that cause the creation of new nodes.

Reply:
> The core issue is that events in the column's network can lead to the creation of new columns,

but it is not responsibility of columns to manage columns; The column should have an output saying "A directory is chosen; a new column should be created". Then the browser, after creating column, should connect with endpoint mentioned and create new column.

Reply:
But this would not eliminate the need for weak pointers, would it?

Reply:
Are you suggesting something like this in the file browsers network to create the cycle?
```
push_column <- any_mut::<AnyFolderContent>();
eval push_column([push_column](content) {
    let new_column = Column::new(&content,&push_column);
    ...
});

Reply:
I opened an issue for it (#1695).

Reply:
> Are you suggesting something like this in the file browsers network to create the cycle?
> 
> ```
> push_column <- any_mut::<AnyFolderContent>();
> eval push_column([push_column](content) {
>     let new_column = Column::new(&content,&push_column);
>     ...
> });
> ```

No, rather something like 
```rust
push_column <- any_mut::<AnyFolderContent>();
eval push_column([push_column](content) {
    let new_column = Column::new(&content);
    push_column <+ new_column.column_should_be_pushed.
});
```
But of course this is a simplification: in fact, the column should not care if another one should be pushed, or maybe columns on the right should be hidden first: it should just have an endpoint "user wants to open directory", and the parent "FileBrowser" should decide if that means pushing new column, or whatever else.

Reply:
I fixed the original concern by using `if let ...` in place of `unwrap`. But I am currently hesitant to refactor the code as proposed, since I cannot discuss it with Adam.